Fix failing smoke tests

The smoke tests are failing:

https://github.com/lephare-photoz/lephare/actions/runs/9560338829

FAILED tests/lephare/test_filter.py::test_filtersvc - ValueError: resize only works on single-segment arrays
FAILED tests/lephare/test_filter.py::test_flt_plot_filter_curve - ValueError: resize only works on single-segment arrays
FAILED tests/lephare/test_sed.py::test_sed_fromfile - ValueError: resize only works on single-segment arrays
FAILED tests/lephare/test_sed.py::test_sed_readwrite - ValueError: resize only works on single-segment arrays

We believ this may be due to a numpy update. We therefore consider temporarily pinning numpy to the previous version and/or updating numpy to be numpy 2.0 compatible.


- [x] I have described the situation in which the bug arose, including what code was executed, information about my environment, and any applicable data others will need to reproduce the problem.
- [x] I have included available evidence of the unexpected behavior (including error messages, screenshots, and/or plots) as well as a descriprion of what I expected instead.
- [x] If I have a solution in mind, I have provided an explanation and/or pseudocode and/or task list.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -21,7 +21,7 @@ dynamic = ["version"]
 requires-python = ">=3.9"
 dependencies = [
     "matplotlib",
-    "numpy<2",
+    "numpy",
     "astropy",
     "scipy",
     "requests",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -111,6 +111,8 @@ select = [
     "D300",
     "D417",
     "D419",
+    # Update numpy to 2.0 
+    "NPY201",
 ]
 
 ignore = [

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -21,7 +21,7 @@ dynamic = ["version"]
 requires-python = ">=3.9"
 dependencies = [
     "matplotlib",
-    "numpy",
+    "numpy<2",
     "astropy",
     "scipy",
     "requests",

--- a/src/lephare/_spec.py
+++ b/src/lephare/_spec.py
@@ -198,5 +198,3 @@ def plotspec(filename):
     ax2.plot(zpdf[0, :], zpdf[1, :] / max(zpdf[1, :]), color="k", linestyle="solid", label="P(z) Bayesian")
     ax2.plot(zpdf[0, :], zpdf[2, :] / max(zpdf[2, :]), color="k", linestyle="dashed", label="P(z) Profile")
     ax2.legend(fontsize="10")
-
-    plt.show()

--- a/src/lib/_bindings.cc
+++ b/src/lib/_bindings.cc
@@ -126,14 +126,14 @@ PYBIND11_MODULE(_lephare, mod) {
       .def_readwrite("lamb_trans", &flt::lamb_trans)
       .def("data", [](const flt &f) {
         int N = f.lamb_trans.size();
-        py::array_t<double> result = py::array_t<double>(N * 2);
+        // Create a 2D array with shape (2, N) (transposed)
+        py::array_t<double> result({2, N});
         py::buffer_info buf = result.request();
-        double *ptr = (double *)buf.ptr;
+        double *ptr = static_cast<double *>(buf.ptr);
         for (size_t i = 0; i < N; i++) {
-          ptr[i] = f.lamb_trans[i].lamb;
-          ptr[N + i] = f.lamb_trans[i].val;
+          ptr[i] = f.lamb_trans[i].lamb;     // First row
+          ptr[N + i] = f.lamb_trans[i].val;  // Second row
         }
-        result.resize({N, N});
         return result;
       });
   mod.def("write_output_filter", &write_output_filter);
@@ -173,14 +173,14 @@ PYBIND11_MODULE(_lephare, mod) {
                                      const string &)>(&SED::writeSED))
       .def("data", [](const SED &f) {
         int N = f.lamb_flux.size();
-        py::array_t<double> result = py::array_t<double>(N * 2);
+        // Create a 2D array with shape (2, N) (transposed)
+        py::array_t<double> result({2, N});
         py::buffer_info buf = result.request();
-        double *ptr = (double *)buf.ptr;
+        double *ptr = static_cast<double *>(buf.ptr);
         for (size_t i = 0; i < N; i++) {
-          ptr[i] = f.lamb_flux[i].lamb;
-          ptr[N + i] = f.lamb_flux[i].val;
+          ptr[i] = f.lamb_flux[i].lamb;     // First row
+          ptr[N + i] = f.lamb_flux[i].val;  // Second row
         }
-        result.resize({N, N});
         return result;
       });
 

--- a/src/lib/_bindings.cc
+++ b/src/lib/_bindings.cc
@@ -126,14 +126,14 @@ PYBIND11_MODULE(_lephare, mod) {
       .def_readwrite("lamb_trans", &flt::lamb_trans)
       .def("data", [](const flt &f) {
         int N = f.lamb_trans.size();
-        // Create a 2D array with shape (N, 2)
-        py::array_t<double> result({N, 2});
+        py::array_t<double> result = py::array_t<double>(N * 2);
         py::buffer_info buf = result.request();
         double *ptr = (double *)buf.ptr;
         for (size_t i = 0; i < N; i++) {
-          ptr[i * 2] = f.lamb_trans[i].lamb;     // First column
-          ptr[i * 2 + 1] = f.lamb_trans[i].val;  // Second column
+          ptr[i] = f.lamb_trans[i].lamb;
+          ptr[N + i] = f.lamb_trans[i].val;
         }
+        result.resize({N, N});
         return result;
       });
   mod.def("write_output_filter", &write_output_filter);
@@ -173,14 +173,14 @@ PYBIND11_MODULE(_lephare, mod) {
                                      const string &)>(&SED::writeSED))
       .def("data", [](const SED &f) {
         int N = f.lamb_flux.size();
-        // Create a 2D array with shape (N, 2)
-        py::array_t<double> result({N, 2});
+        py::array_t<double> result = py::array_t<double>(N * 2);
         py::buffer_info buf = result.request();
         double *ptr = (double *)buf.ptr;
         for (size_t i = 0; i < N; i++) {
-          ptr[i * 2] = f.lamb_flux[i].lamb;     // First column
-          ptr[i * 2 + 1] = f.lamb_flux[i].val;  // Second column
+          ptr[i] = f.lamb_flux[i].lamb;
+          ptr[N + i] = f.lamb_flux[i].val;
         }
+        result.resize({N, N});
         return result;
       });
 

--- a/src/lib/_bindings.cc
+++ b/src/lib/_bindings.cc
@@ -126,14 +126,14 @@ PYBIND11_MODULE(_lephare, mod) {
       .def_readwrite("lamb_trans", &flt::lamb_trans)
       .def("data", [](const flt &f) {
         int N = f.lamb_trans.size();
-        py::array_t<double> result = py::array_t<double>(N * 2);
+        // Create a 2D array with shape (N, 2)
+        py::array_t<double> result({N, 2});
         py::buffer_info buf = result.request();
         double *ptr = (double *)buf.ptr;
         for (size_t i = 0; i < N; i++) {
-          ptr[i] = f.lamb_trans[i].lamb;
-          ptr[N + i] = f.lamb_trans[i].val;
+          ptr[i * 2] = f.lamb_trans[i].lamb;     // First column
+          ptr[i * 2 + 1] = f.lamb_trans[i].val;  // Second column
         }
-        result.resize({N, N});
         return result;
       });
   mod.def("write_output_filter", &write_output_filter);
@@ -173,14 +173,14 @@ PYBIND11_MODULE(_lephare, mod) {
                                      const string &)>(&SED::writeSED))
       .def("data", [](const SED &f) {
         int N = f.lamb_flux.size();
-        py::array_t<double> result = py::array_t<double>(N * 2);
+        // Create a 2D array with shape (N, 2)
+        py::array_t<double> result({N, 2});
         py::buffer_info buf = result.request();
         double *ptr = (double *)buf.ptr;
         for (size_t i = 0; i < N; i++) {
-          ptr[i] = f.lamb_flux[i].lamb;
-          ptr[N + i] = f.lamb_flux[i].val;
+          ptr[i * 2] = f.lamb_flux[i].lamb;     // First column
+          ptr[i * 2 + 1] = f.lamb_flux[i].val;  // Second column
         }
-        result.resize({N, N});
         return result;
       });
 

--- a/tests/lephare/test_pdf.py
+++ b/tests/lephare/test_pdf.py
@@ -15,7 +15,7 @@ def test_pdf():
 def test_set_yvals():
     """Test to make sure that chi2 and vPDF are set correctly."""
     test_pdf = lp.PDF(0, 3, 10)
-    yvals = np.linspace(0, 3, 10)
+    yvals = np.linspace(0.001, 3, 10)
     test_pdf.setYvals(yvals, is_chi2=True)
     np.testing.assert_array_equal(test_pdf.chi2, yvals)
     np.testing.assert_array_almost_equal(test_pdf.vPDF, np.exp(-0.5 * yvals))
@@ -28,7 +28,7 @@ def test_set_yvals():
 def test_plot():
     """Simple test to exercise the plot method."""
     test_pdf = lp.PDF(0, 3, 10)
-    yvals = np.linspace(0, 3, 10)
+    yvals = np.linspace(0.001, 3, 10)
     test_pdf.setYvals(yvals, is_chi2=True)
     test_pdf.plot(chi2=True)
     test_pdf.plot(chi2=False)

